aten/src/ATen/native/transformers/cuda/sdp_utils.h

```diff
@@ -0,0 +1,241 @@
+#pragma once
+
+#include <ATen/Context.h>
+#include <ATen/TensorUtils.h>
+#include <ATen/core/Tensor.h>
+#include <ATen/detail/CUDAHooksInterface.h>
+#include <ATen/native/DispatchStub.h>
+#include <c10/core/ScalarType.h>
+#include <c10/util/env.h>
+#include <c10/util/irange.h>
+
+#include <functional>
+#include <unordered_set>
+#include <vector>
+#include <ATen/cuda/CUDAContext.h>
+
+namespace sdp {
+
+struct sdp_params {
+  const at::Tensor& query;
+  const at::Tensor& key;
+  const at::Tensor& value;
+  bool has_attn_mask;
+  double dropout;
+  bool need_attn_weights;
+  bool is_causal;
+};
+
+enum class SDPBackend {flash_attention, math, error};
+
+//  Define gate functions that determine if a flash kernel can be ran
+inline bool check_gradients(sdp_params params, bool debug) {
+  if (params.query.requires_grad() || params.key.requires_grad() ||
+      params.value.requires_grad()) {
+    TORCH_CHECK(
+        debug,
+        "Query, Key, or Value require's gradient. Query: ",
+        params.query.requires_grad(),
+        " Key: ",
+        params.key.requires_grad(),
+        " Value: ",
+        params.value.requires_grad());
+    return false;
+  }
+  return true;
+}
+
+inline bool compiled_with_flash_kernels(sdp_params params, bool debug) {
+#ifndef USE_FLASH_ATTENTION
```

Comment:
I think it makes more sense to guard `return False` within `use_flash_attention`. That'll give the compiler the chance to just entirely ignore the function instead of running a for loop on std::functions.